File: src/main/java/frc/team5190/robot/drive/DriveTrainSubsystem.java
package frc.team5190.robot.drive;

import com.ctre.phoenix.drive.DiffDrive;
import com.ctre.phoenix.drive.DriveMode;
import com.ctre.phoenix.motorcontrol.ControlMode;
import com.ctre.phoenix.motorcontrol.FeedbackDevice;
import com.ctre.phoenix.motorcontrol.can.TalonSRX;
import com.kauailabs.navx.frc.AHRS;
import edu.wpi.first.wpilibj.GenericHID;
import edu.wpi.first.wpilibj.SPI;
import edu.wpi.first.wpilibj.buttons.JoystickButton;
import edu.wpi.first.wpilibj.command.Subsystem;
import frc.team5190.robot.Constants;
import frc.team5190.robot.Maths;
import frc.team5190.robot.OI;

import javax.naming.ldap.Control;

@SuppressWarnings("WeakerAccess")
public class DriveTrainSubsystem extends Subsystem {

    public TalonSRX frontLeft;
    public TalonSRX frontRight;
    public TalonSRX rearLeft;
    public TalonSRX rearRight;

    public AHRS navX;

    private DiffDrive driveBase;

    public DriveTrainSubsystem() {
        frontLeft   = new TalonSRX(Constants.INSTANCE.getFrontLeftMotorVal());
        frontRight  = new TalonSRX(Constants.INSTANCE.getFrontRightMotorVal());
        rearLeft    = new TalonSRX(Constants.INSTANCE.getRearLeftMotorVal());
        rearRight   = new TalonSRX(Constants.INSTANCE.getRearRightMotorVal());

        frontLeft.set(ControlMode.Velocity, 0);
        frontLeft.configSelectedFeedbackSensor(FeedbackDevice.QuadEncoder, 0, 0);
        frontLeft.setSensorPhase(true);

        frontLeft.config_kP(0, 0.6, 0);
        frontLeft.config_kF(0, Maths.calculateFGain(1, 18.3f, 2, 1440), 0);

        rearLeft.follow(frontLeft);

        frontRight.set(ControlMode.Velocity, 0);
        frontRight.configSelectedFeedbackSensor(FeedbackDevice.QuadEncoder, 0, 0);
        frontRight.setSensorPhase(true);

        frontRight.config_kP(0, 0.6, 0);
        frontLeft.config_kF(0, Maths.calculateFGain(1, 18.3f, 2, 1440), 0);

        rearRight.follow(frontRight);

        navX = new AHRS(SPI.Port.kMXP);

        driveBase = new DiffDrive(frontLeft, frontRight);

//        JoystickButton aButton = new JoystickButton(OI.INSTANCE.getXbox(), 1);
//        aButton.whenPressed(new FootCommand());
    }

    @Override
    protected void initDefaultCommand() {
        this.setDefaultCommand(new DriveCommand());
    }

    public void testDrive() {
        falconTankDrive(OI.INSTANCE.getXbox().getY(GenericHID.Hand.kRight), -OI.INSTANCE.getXbox().getY(GenericHID.Hand.kLeft));
    }

    public void turn(double curve) {
        driveBase.set(DriveMode.PercentOutput, 0, curve);
    }


    public void falconTankDrive(double leftValue, double rightValue)
    {
        if (leftValue >= 0.0) {
            leftValue = leftValue * leftValue;
        }
        else {
            leftValue = -(leftValue * leftValue);
        }
        if (rightValue >= 0.0) {
            rightValue = rightValue * rightValue;
        }
        else {
            rightValue = -(rightValue * rightValue);
        }

        leftValue *= 1049;
        rightValue *= 1049;

        frontLeft.set(ControlMode.Velocity, leftValue);
        frontRight.set(ControlMode.Velocity, rightValue);

    }
}

File: src/main/kotlin/frc/team5190/robot/drive/DriveTrainSubsystem.kt
package frc.team5190.robot.drive

import com.ctre.phoenix.motorcontrol.ControlMode
import com.ctre.phoenix.motorcontrol.NeutralMode
import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX
import edu.wpi.first.wpilibj.GenericHID
import edu.wpi.first.wpilibj.command.Subsystem
import edu.wpi.first.wpilibj.drive.DifferentialDrive
import frc.team5190.robot.MotorIds
import frc.team5190.robot.OI
import kotlin.math.absoluteValue

enum class DriveMode {
    TANK,
    CURVE
}

class DriveTrainSubsystem : Subsystem() {

    var controlMode = DriveMode.TANK
        set(value) {
            println("Changing DriveMode from $field to $value")
            field = value
        }

    private val frontLeft = WPI_TalonSRX(MotorIds.FRONT_LEFT_VAL)
    private val frontRight = WPI_TalonSRX(MotorIds.FRONT_RIGHT_VAL)

    init {
        println("Init Drive")

        WPI_TalonSRX(MotorIds.REAR_LEFT_VAL).follow(frontLeft)
        WPI_TalonSRX(MotorIds.REAR_RIGHT_VAL).follow(frontRight)

        listOf(frontLeft, frontRight).forEach { it.setNeutralMode(NeutralMode.Brake) }
        println("Drive Done")
    }

    override fun initDefaultCommand() {
        this.defaultCommand = DriveCommand()
    }

    fun drive() {
        OI.xbox.getY(GenericHID.Hand.kLeft).takeIf { it.absoluteValue > 0.2 }?.let { frontRight.set(it) }
        OI.xbox.getY(GenericHID.Hand.kRight).takeIf { it.absoluteValue > 0.2 }?.let { frontLeft.set(-it) }
        if (OI.xbox.backButtonPressed) controlMode = DriveMode.TANK
        if (OI.xbox.startButtonPressed) controlMode = DriveMode.CURVE
    }
}

